File: Mac/ES_R_DevKit_Mac/MEME_Academic/DecEnc.m
//
//  DecEnc.m
//  DecEnc
//
//  Created by GitHub Copilot on 2026/05/15.
//

#import "DecEnc.h"

@implementation DecEnc

+ (void)Encode:(uint8_t *)buf {
    const uint8_t key[18] = {0x39,0xCC,0x6D,0xAB,0x9E,0x07,0x1A,0xDE,0x67,0x49,0x71,0x9A,0x5B,0x69,0x0F,0x17,0xC9,0xB1};
    const size_t keyLen = sizeof(key) / sizeof(key[0]);
    for (size_t i = 0; i < keyLen; ++i) {
        uint8_t v = buf[i];
        uint8_t enc = (uint8_t)(0xFF & ((v ^ key[i]) + (uint8_t)i));
        buf[i] = enc;
    }
}

+ (void)Decode:(uint8_t *)buf {
    const uint8_t key[18] = {0x39,0xCC,0x6D,0xAB,0x9E,0x07,0x1A,0xDE,0x67,0x49,0x71,0x9A,0x5B,0x69,0x0F,0x17,0xC9,0xB1};
    const size_t keyLen = sizeof(key) / sizeof(key[0]);
    for (size_t i = 0; i < keyLen; ++i) {
        uint8_t v = buf[i];
        uint8_t dec = (uint8_t)(0xFF & (((uint8_t)(v - (uint8_t)i)) ^ key[i]));
        buf[i] = dec;
    }
}

@end

File: windows/ES_R_DevKit_Windows/MEMELib_Academic/DecEnc.cs
namespace MEMELib_Academic;

/// <summary>
/// ES_R の 20 byte パケットに掛かっている簡易難読化。先頭 2 byte
/// (長さ・オペコード)は素通しで、残り 18 byte を固定鍵で変換する。
/// Mac 版 <c>DecEnc.swift</c> / Android 版 <c>DataEncryption</c> と同じアルゴリズム。
/// </summary>
public static class DecEnc
{
    private static readonly byte[] Key =
    [
        0x39, 0xCC, 0x6D, 0xAB, 0x9E, 0x07, 0x1A, 0xDE, 0x67,
        0x49, 0x71, 0x9A, 0x5B, 0x69, 0x0F, 0x17, 0xC9, 0xB1,
    ];

    /// <summary>鍵長。ヘッダ 2 byte を除いたパケット本体の長さと一致する。</summary>
    public static int KeyLength => Key.Length;

    /// <summary>送信前の変換。buf は 20 byte 必要。</summary>
    public static void Encode(Span<byte> buf)
    {
        if (buf.Length < 2 + Key.Length)
        {
            throw new ArgumentException($"buffer must be at least {2 + Key.Length} bytes", nameof(buf));
        }

        for (int i = 0; i < Key.Length; i++)
        {
            buf[2 + i] = unchecked((byte)((buf[2 + i] ^ Key[i]) + i));
        }
    }

    /// <summary>受信後の復号。<see cref="Encode"/> の逆変換。</summary>
    public static void Decode(Span<byte> buf)
    {
        if (buf.Length < 2 + Key.Length)
        {
            throw new ArgumentException($"buffer must be at least {2 + Key.Length} bytes", nameof(buf));
        }

        for (int i = 0; i < Key.Length; i++)
        {
            buf[2 + i] = unchecked((byte)((byte)(buf[2 + i] - i) ^ Key[i]));
        }
    }
}
